dash/include/dash/Coarray.h

```diff
@@ -0,0 +1,771 @@
+                              pattern_type>;
+};
+
+template<
+  typename element_type,
+  typename pattern_type>
+struct __get_local_type<element_type, pattern_type, 1> {
+  using type = element_type &;
+};
+
+template<typename element_type>
+struct __get_ref_type {
+// GAR needs conversion operator
+//  using type = GlobAsyncRef<element_type>;
+  using type = GlobRef<element_type>;
+};
+
+/**
+ * atomics cannot be accessed asynchronously
+ */
+template<typename element_type>
+struct __get_ref_type<dash::Atomic<element_type>> {
+  using type = GlobRef<dash::Atomic<element_type>>;
+};
+
+template<typename element_type>
+struct __get_const_ref_type {
+  using type = GlobAsyncRef<const element_type>;
+};
+
+/**
+ * atomics cannot be accessed asynchronously
+ */
+template<typename element_type>
+struct __get_const_ref_type<dash::Atomic<element_type>> {
+  using type = GlobRef<dash::Atomic<const element_type>>;
+};
+
+} // namespace detail
+
+/**
+ * helper to create a coarray pattern for coarrays where the local size of
+ * each unit is equal (symmetric allocation).
+ */
+template<
+  typename T,
+  typename IndexType = dash::default_index_t>
+struct make_coarray_symmetric_pattern {
+  using type = BlockPattern<
+                std::rank<
+                  typename dash::remove_atomic<T>::type>::value+1,
+                ROW_MAJOR,
+                IndexType>;
+};
+
+} // namespace coarray
+
+/**
+ * A fortran style coarray.
+ * 
+ * \ingroup DashCoarrayConcept
+ * 
+ * Interface of the Coarray for scalar and array types showing local
+ * and global accesses:
+ * \code
+ * dash::Coarray<int> i;          // scalar Coarray
+ * dash::Coarray<int[10][20]> x;  // 2D-Coarray
+ * dash::Coarray<int[][20]> y(n); // one open dim,
+ *                                // set at runtime in ctor
+ * 
+ * // access syntax
+ * i(unit) = value; // global access
+ * i       = value; // local access
+ * 
+ * x(unit)[idx1][idx2] = value; // global access
+ * x[idx1][idx2]       = value; // local access
+ * \endcode
+ *
+ */
+template<
+  typename T,
+  typename IndexType = dash::default_index_t,
+  class PatternType  = typename coarray::make_coarray_symmetric_pattern<T,IndexType>::type >
+class Coarray {
+private:
+
+  using self_t          = Coarray<T, IndexType, PatternType>;
+  
+  /**
+   * _element_type has no extent and is wrapped with \cdash::Atomic, if coarray
+   * was declared with \cdash::Atomic<T>
+   */
+  using _element_type   = typename std::remove_all_extents<T>::type;
+  using _underl_type    = typename dash::remove_atomic<_element_type>::type;
+  
+  using _rank           = std::integral_constant<int,
+                            std::rank<T>::value>;
+  
+  using _size_type      = typename std::make_unsigned<IndexType>::type;
+  using _sspec_type     = SizeSpec<_rank::value+1, _size_type>;
+  using _pattern_type   = PatternType;
+  
+  using _storage_type   = Matrix<_element_type, _rank::value+1, IndexType, _pattern_type>;
```

Comment:
Just a thought I had: could we use `dash::Array` as underlying storage type for scalar CoArrays?

Reply:
Theoretically possible, but currently not fully implemented as some types mismatch. For the paper I tested it, so see it as future work ;)